### ISSUE TITLE
data race

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#197]
Saw a data race on a CI run ... (at corp dd6c460e5a51322235118476e20ac638b9242b38)

```
2020-03-19T20:43:20.9029876Z ==================
2020-03-19T20:43:20.9030017Z WARNING: DATA RACE
2020-03-19T20:43:20.9030178Z Write at 0x00c00006c9b0 by goroutine 92:
2020-03-19T20:43:20.9030553Z   github.com/tailscale/wireguard-go/conn.(*NativeEndpoint).UpdateDst()
2020-03-19T20:43:20.9031083Z       /home/runner/work/corp/corp/wireguard-go/conn/conn_linux.go:102 +0x496
2020-03-19T20:43:20.9031593Z   github.com/tailscale/wireguard-go/device.(*Peer).SetEndpointAddress()
2020-03-19T20:43:20.9031931Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:335 +0x18f
2020-03-19T20:43:20.9032268Z   github.com/tailscale/wireguard-go/device.(*Device).RoutineHandshake()
2020-03-19T20:43:20.9032605Z       /home/runner/work/corp/corp/wireguard-go/device/receive.go:453 +0xea8
2020-03-19T20:43:20.9032700Z 
2020-03-19T20:43:20.9032859Z Previous read at 0x00c00006c9b0 by goroutine 101:
2020-03-19T20:43:20.9032996Z   reflect.Value.Bool()
2020-03-19T20:43:20.9033183Z       /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:267 +0x4e
2020-03-19T20:43:20.9033323Z   fmt.(*pp).printValue()
2020-03-19T20:43:20.9033505Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:747 +0x3464
2020-03-19T20:43:20.9033648Z   fmt.(*pp).printValue()
2020-03-19T20:43:20.9033827Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:810 +0x283f
2020-03-19T20:43:20.9033967Z   fmt.(*pp).printValue()
2020-03-19T20:43:20.9034149Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:880 +0x25da
2020-03-19T20:43:20.9034287Z   fmt.(*pp).printArg()
2020-03-19T20:43:20.9034465Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:716 +0x2ee
2020-03-19T20:43:20.9034603Z   fmt.(*pp).doPrintf()
2020-03-19T20:43:20.9034784Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:1030 +0x312
2020-03-19T20:43:20.9034919Z   fmt.Sprintf()
2020-03-19T20:43:20.9035097Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:219 +0x73
2020-03-19T20:43:20.9035237Z   log.(*Logger).Printf()
2020-03-19T20:43:20.9035415Z       /opt/hostedtoolcache/go/1.13.7/x64/src/log/log.go:179 +0x64
2020-03-19T20:43:20.9035781Z   github.com/tailscale/wireguard-go/device.(*Peer).SendHandshakeInitiation()
2020-03-19T20:43:20.9036125Z       /home/runner/work/corp/corp/wireguard-go/device/send.go:152 +0x2fe
2020-03-19T20:43:20.9036451Z   github.com/tailscale/wireguard-go/device.(*Peer).RoutineNonce()
2020-03-19T20:43:20.9036786Z       /home/runner/work/corp/corp/wireguard-go/device/send.go:475 +0x296
2020-03-19T20:43:20.9036855Z 
2020-03-19T20:43:20.9036993Z Goroutine 92 (running) created at:
2020-03-19T20:43:20.9037289Z   github.com/tailscale/wireguard-go/device.NewDevice()
2020-03-19T20:43:20.9037623Z       /home/runner/work/corp/corp/wireguard-go/device/device.go:401 +0x5c7
2020-03-19T20:43:20.9038079Z   github.com/tailscale/wireguard-go/device.TestSimultaneousHandshake()
2020-03-19T20:43:20.9038614Z       /home/runner/work/corp/corp/wireguard-go/device/device_test.go:117 +0x7ca
2020-03-19T20:43:20.9038758Z   testing.tRunner()
2020-03-19T20:43:20.9038947Z       /opt/hostedtoolcache/go/1.13.7/x64/src/testing/testing.go:909 +0x199
2020-03-19T20:43:20.9039033Z 
2020-03-19T20:43:20.9039176Z Goroutine 101 (running) created at:
2020-03-19T20:43:20.9039563Z   github.com/tailscale/wireguard-go/device.(*Peer).Start()
2020-03-19T20:43:20.9039992Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:238 +0x4db
2020-03-19T20:43:20.9040324Z   github.com/tailscale/wireguard-go/device.(*Device).NewPeer()
2020-03-19T20:43:20.9040667Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:143 +0x838
2020-03-19T20:43:20.9041010Z   github.com/tailscale/wireguard-go/device.(*Device).IpcSetOperation()
2020-03-19T20:43:20.9041358Z       /home/runner/work/corp/corp/wireguard-go/device/uapi.go:247 +0x1743
2020-03-19T20:43:20.9041708Z   github.com/tailscale/wireguard-go/device.TestSimultaneousHandshake()
2020-03-19T20:43:20.9042066Z       /home/runner/work/corp/corp/wireguard-go/device/device_test.go:122 +0x8fa
2020-03-19T20:43:20.9042552Z   testing.tRunner()
2020-03-19T20:43:20.9042743Z       /opt/hostedtoolcache/go/1.13.7/x64/src/testing/testing.go:909 +0x199
2020-03-19T20:43:20.9042883Z ==================
2020-03-19T20:43:20.9043008Z ==================
2020-03-19T20:43:20.9043139Z WARNING: DATA RACE
2020-03-19T20:43:20.9043519Z Write at 0x00c00006c968 by goroutine 92:
2020-03-19T20:43:20.9044111Z   github.com/tailscale/wireguard-go/conn.(*NativeEndpoint).UpdateDst()
2020-03-19T20:43:20.9044500Z       /home/runner/work/corp/corp/wireguard-go/conn/conn_linux.go:103 +0x4dd
2020-03-19T20:43:20.9044879Z   github.com/tailscale/wireguard-go/device.(*Peer).SetEndpointAddress()
2020-03-19T20:43:20.9045258Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:335 +0x18f
2020-03-19T20:43:20.9045640Z   github.com/tailscale/wireguard-go/device.(*Device).RoutineHandshake()
2020-03-19T20:43:20.9046022Z       /home/runner/work/corp/corp/wireguard-go/device/receive.go:453 +0xea8
2020-03-19T20:43:20.9046268Z 
2020-03-19T20:43:20.9046436Z Previous read at 0x00c00006c968 by goroutine 101:
2020-03-19T20:43:20.9046585Z   reflect.typedmemmove()
2020-03-19T20:43:20.9046945Z       /opt/hostedtoolcache/go/1.13.7/x64/src/runtime/mbarrier.go:177 +0x0
2020-03-19T20:43:20.9047252Z   reflect.packEface()
2020-03-19T20:43:20.9047602Z       /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:119 +0x103
2020-03-19T20:43:20.9047746Z   reflect.valueInterface()
2020-03-19T20:43:20.9047932Z       /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:1033 +0x16f
2020-03-19T20:43:20.9048072Z   fmt.(*pp).printValue()
2020-03-19T20:43:20.9048256Z       /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:1003 +0x38f7
2020-03-19T20:43:20.9048397Z   fmt.(*pp).printValue()
2020-03-19T20:43:20.9048577Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:880 +0x25da
2020-03-19T20:43:20.9048722Z   fmt.(*pp).printArg()
2020-03-19T20:43:20.9048900Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:716 +0x2ee
2020-03-19T20:43:20.9049041Z   fmt.(*pp).doPrintf()
2020-03-19T20:43:20.9049221Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:1030 +0x312
2020-03-19T20:43:20.9049352Z   fmt.Sprintf()
2020-03-19T20:43:20.9049527Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:219 +0x73
2020-03-19T20:43:20.9049669Z   log.(*Logger).Printf()
2020-03-19T20:43:20.9049846Z       /opt/hostedtoolcache/go/1.13.7/x64/src/log/log.go:179 +0x64
2020-03-19T20:43:20.9050211Z   github.com/tailscale/wireguard-go/device.(*Peer).SendHandshakeInitiation()
2020-03-19T20:43:20.9050551Z       /home/runner/work/corp/corp/wireguard-go/device/send.go:152 +0x2fe
2020-03-19T20:43:20.9050873Z   github.com/tailscale/wireguard-go/device.(*Peer).RoutineNonce()
2020-03-19T20:43:20.9051602Z       /home/runner/work/corp/corp/wireguard-go/device/send.go:475 +0x296
2020-03-19T20:43:20.9051780Z 
2020-03-19T20:43:20.9051923Z Goroutine 92 (running) created at:
2020-03-19T20:43:20.9052278Z   github.com/tailscale/wireguard-go/device.NewDevice()
2020-03-19T20:43:20.9052632Z       /home/runner/work/corp/corp/wireguard-go/device/device.go:401 +0x5c7
2020-03-19T20:43:20.9052980Z   github.com/tailscale/wireguard-go/device.TestSimultaneousHandshake()
2020-03-19T20:43:20.9053415Z       /home/runner/work/corp/corp/wireguard-go/device/device_test.go:117 +0x7ca
2020-03-19T20:43:20.9053568Z   testing.tRunner()
2020-03-19T20:43:20.9053828Z       /opt/hostedtoolcache/go/1.13.7/x64/src/testing/testing.go:909 +0x199
2020-03-19T20:43:20.9053899Z 
2020-03-19T20:43:20.9054041Z Goroutine 101 (running) created at:
2020-03-19T20:43:20.9054545Z   github.com/tailscale/wireguard-go/device.(*Peer).Start()
2020-03-19T20:43:20.9054880Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:238 +0x4db
2020-03-19T20:43:20.9055201Z   github.com/tailscale/wireguard-go/device.(*Device).NewPeer()
2020-03-19T20:43:20.9055545Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:143 +0x838
2020-03-19T20:43:20.9055878Z   github.com/tailscale/wireguard-go/device.(*Device).IpcSetOperation()
2020-03-19T20:43:20.9056218Z       /home/runner/work/corp/corp/wireguard-go/device/uapi.go:247 +0x1743
2020-03-19T20:43:20.9056551Z   github.com/tailscale/wireguard-go/device.TestSimultaneousHandshake()
2020-03-19T20:43:20.9056905Z       /home/runner/work/corp/corp/wireguard-go/device/device_test.go:122 +0x8fa
2020-03-19T20:43:20.9057222Z   testing.tRunner()
2020-03-19T20:43:20.9057413Z       /opt/hostedtoolcache/go/1.13.7/x64/src/testing/testing.go:909 +0x199
2020-03-19T20:43:20.9057555Z ==================
2020-03-19T20:43:20.9057681Z ==================
2020-03-19T20:43:20.9057806Z WARNING: DATA RACE
2020-03-19T20:43:20.9058139Z Write at 0x00c00006c9a0 by goroutine 92:
2020-03-19T20:43:20.9059135Z   github.com/tailscale/wireguard-go/conn.(*NativeEndpoint).ClearSrc()
2020-03-19T20:43:20.9059524Z       /home/runner/work/corp/corp/wireguard-go/conn/conn_linux.go:323 +0x4b
2020-03-19T20:43:20.9059876Z   github.com/tailscale/wireguard-go/conn.(*NativeEndpoint).UpdateDst()
2020-03-19T20:43:20.9060238Z       /home/runner/work/corp/corp/wireguard-go/conn/conn_linux.go:105 +0x548
2020-03-19T20:43:20.9060593Z   github.com/tailscale/wireguard-go/device.(*Peer).SetEndpointAddress()
2020-03-19T20:43:20.9060951Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:335 +0x18f
2020-03-19T20:43:20.9061633Z   github.com/tailscale/wireguard-go/device.(*Device).RoutineHandshake()
2020-03-19T20:43:20.9061995Z       /home/runner/work/corp/corp/wireguard-go/device/receive.go:453 +0xea8
2020-03-19T20:43:20.9062086Z 
2020-03-19T20:43:20.9062233Z Previous read at 0x00c00006c9a0 by goroutine 101:
2020-03-19T20:43:20.9062380Z   reflect.typedmemmove()
2020-03-19T20:43:20.9062577Z       /opt/hostedtoolcache/go/1.13.7/x64/src/runtime/mbarrier.go:177 +0x0
2020-03-19T20:43:20.9062726Z   reflect.packEface()
2020-03-19T20:43:20.9062924Z       /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:119 +0x103
2020-03-19T20:43:20.9063074Z   reflect.valueInterface()
2020-03-19T20:43:20.9063270Z       /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:1033 +0x16f
2020-03-19T20:43:20.9063422Z   fmt.(*pp).printValue()
2020-03-19T20:43:20.9063616Z       /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:1003 +0x38f7
2020-03-19T20:43:20.9063766Z   fmt.(*pp).printValue()
2020-03-19T20:43:20.9063957Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:880 +0x25da
2020-03-19T20:43:20.9064102Z   fmt.(*pp).printArg()
2020-03-19T20:43:20.9064289Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:716 +0x2ee
2020-03-19T20:43:20.9064437Z   fmt.(*pp).doPrintf()
2020-03-19T20:43:20.9064626Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:1030 +0x312
2020-03-19T20:43:20.9064917Z   fmt.Sprintf()
2020-03-19T20:43:20.9065097Z       /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:219 +0x73
2020-03-19T20:43:20.9065644Z   log.(*Logger).Printf()
2020-03-19T20:43:20.9065829Z       /opt/hostedtoolcache/go/1.13.7/x64/src/log/log.go:179 +0x64
2020-03-19T20:43:20.9066425Z   github.com/tailscale/wireguard-go/device.(*Peer).SendHandshakeInitiation()
2020-03-19T20:43:20.9066792Z       /home/runner/work/corp/corp/wireguard-go/device/send.go:152 +0x2fe
2020-03-19T20:43:20.9067134Z   github.com/tailscale/wireguard-go/device.(*Peer).RoutineNonce()
2020-03-19T20:43:20.9067759Z       /home/runner/work/corp/corp/wireguard-go/device/send.go:475 +0x296
2020-03-19T20:43:20.9067858Z 
2020-03-19T20:43:20.9068011Z Goroutine 92 (running) created at:
2020-03-19T20:43:20.9068358Z   github.com/tailscale/wireguard-go/device.NewDevice()
2020-03-19T20:43:20.9068726Z       /home/runner/work/corp/corp/wireguard-go/device/device.go:401 +0x5c7
2020-03-19T20:43:20.9069086Z   github.com/tailscale/wireguard-go/device.TestSimultaneousHandshake()
2020-03-19T20:43:20.9069801Z       /home/runner/work/corp/corp/wireguard-go/device/device_test.go:117 +0x7ca
2020-03-19T20:43:20.9069961Z   testing.tRunner()
2020-03-19T20:43:20.9070160Z       /opt/hostedtoolcache/go/1.13.7/x64/src/testing/testing.go:909 +0x199
2020-03-19T20:43:20.9070235Z 
2020-03-19T20:43:20.9070719Z Goroutine 101 (running) created at:
2020-03-19T20:43:20.9071061Z   github.com/tailscale/wireguard-go/device.(*Peer).Start()
2020-03-19T20:43:20.9071425Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:238 +0x4db
2020-03-19T20:43:20.9071951Z   github.com/tailscale/wireguard-go/device.(*Device).NewPeer()
2020-03-19T20:43:20.9072331Z       /home/runner/work/corp/corp/wireguard-go/device/peer.go:143 +0x838
2020-03-19T20:43:20.9072704Z   github.com/tailscale/wireguard-go/device.(*Device).IpcSetOperation()
2020-03-19T20:43:20.9073394Z       /home/runner/work/corp/corp/wireguard-go/device/uapi.go:247 +0x1743
2020-03-19T20:43:20.9073901Z   github.com/tailscale/wireguard-go/device.TestSimultaneousHandshake()
2020-03-19T20:43:20.9074258Z       /home/runner/work/corp/corp/wireguard-go/device/device_test.go:122 +0x8fa
2020-03-19T20:43:20.9074404Z   testing.tRunner()
2020-03-19T20:43:20.9074592Z       /opt/hostedtoolcache/go/1.13.7/x64/src/testing/testing.go:909 +0x199
2020-03-19T20:43:20.9074731Z ==================
```

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1171685499906294/1171686878754957) by [Unito](https://www.unito.io/learn-more)

**Comment by danderson** [comment]
#225 is another facet of the same problem, which is that NativeEndpoint is super unsafe and racey, and needs a ground-up rewrite.

**Comment by danderson** [comment]
Believe this is fixed by tailscale/wireguard-go@23aababa2084b7794e4b75ebbafe92e9d4b94404 also.
